### DISCUSSION TITLE
Pyinstaller Executable says module not found.

### DISCUSSION THREAD

**Discussion by hakunaMatataHub** [pyinstaller/pyinstaller#6948] [Help]
I have a python script for which I want to create execuable binaries using pyinstaller. It imports a packgae called 'rclpy'. I am able to run the script using python3 'script.py' but when I create a binary using pyinstaller it says modeule not found.

While running pyinstaller it does not show any errors.

Here is the error :

Traceback (most recent call last):
  File "operator_dashboard.py", line 22, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "rclpy/node.py", line 41, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "rclpy/client.py", line 22, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "rclpy/impl/implementation_singleton.py", line 31, in <module>
  File "rclpy/impl/__init__.py", line 28, in _import
  File "importlib/__init__.py", line 127, in import_module
ModuleNotFoundError: No module named 'rclpy._rclpy'
The C extension '/home/kewal/mowito_ws/arm_ws/src/order_control/scripts/output/operator_dashboard/rclpy/_rclpy.cpython-38-x86_64-linux-gnu.so' isn't present on the system. Please refer to 'https://index.ros.org/doc/ros2/Troubleshooting/Installation-Troubleshooting/#import-failing-without-library-present-on-the-system' for possible solutions
[166554] Failed to execute script 'operator_dashboard' due to unhandled exception!

There is no directory created for rclpy inside dist either. I am able to locate the pyhton package at /opt/ros/foxy/lib/python3.8/site-packages in my pc.

Please help. Also if I remove all dependecy from this particular package it just works fine.

**Comment by rokm** [comment]
Does adding `--hiddenimport rclpy._rclpy` to your PyInstaller command help?

**Reply by hakunaMatataHub** [reply]
Hey this helped but I guess I have some more hidden imports 
The commend now looks like  **pyinstaller --hiddenimport  rclpy._rclpy_action --hiddenimport rclpy._rclpy --hiddenimport rclpy._rclpy_logging --hiddenimport rclpy._rclpy_signal_handler operator_dashboard.py**
And even after all these hidden import I still have a 'no module find error'. How can I know all hidden imports at once?

**Reply by hakunaMatataHub** [reply]
pyinstaller --hiddenimport  rclpy._rclpy_action --hiddenimport rclpy._rclpy --hiddenimport rclpy._rclpy_logging --hiddenimport rclpy._rclpy_signal_handler --hiddenimport rclpy._rclpy_handle --hiddenimport rclpy._rclpy_pycapsule --hidden-import rcl_interfaces.rcl_interfaces_s__rosidl_typesupport_c --hidden-import builtin_interfaces.builtin_interfaces_s__rosidl_typesupport_c --hiddenimport ur5_msgs.ur5_msgs_s__rosidl_typesupport_c --hiddenimport geometry_msgs.geometry_msgs_s__rosidl_typesupport_c --hiddenimport std_srvs.std_srvs_s__rosidl_typesupport_c operator_dashboard.py

This worked. Also I had to make a directory in dist folder and copy one file from sitet-packages to it.

**Comment by hakunaMatataHub** [comment]
I want to also change its icon

This is the my command and the logs
**pyinstaller --hiddenimport  rclpy._rclpy_action --hiddenimport rclpy._rclpy --hiddenimport rclpy._rclpy_logging --hiddenimport rclpy._rclpy_signal_handler --hiddenimport rclpy._rclpy_handle --hiddenimport rclpy._rclpy_pycapsule --hidden-import rcl_interfaces.rcl_interfaces_s__rosidl_typesupport_c --hidden-import builtin_interfaces.builtin_interfaces_s__rosidl_typesupport_c --hiddenimport ur5_msgs.ur5_msgs_s__rosidl_typesupport_c --hiddenimport geometry_msgs.geometry_msgs_s__rosidl_typesupport_c --hiddenimport std_srvs.std_srvs_s__rosidl_typesupport_c -i /home/kewal/Downloads/iconfinder-icon.svg operator_dashboard.py** 
26 INFO: PyInstaller: 5.2
26 INFO: Python: 3.8.10
31 INFO: Platform: Linux-5.13.0-52-generic-x86_64-with-glibc2.29
32 INFO: wrote /home/kewal/mowito_ws/arm_ws/src/order_control/scripts/operator_dashboard.spec
33 INFO: UPX is not available.
34 INFO: Extending PYTHONPATH with paths
['/home/kewal/mowito_ws/arm_ws/src/order_control/scripts']
[INFO   ] [Logger      ] Record log in /home/kewal/.kivy/logs/kivy_22-07-16_14.txt
61 INFO: [Logger      ] Record log in /home/kewal/.kivy/logs/kivy_22-07-16_14.txt
[INFO   ] [Kivy        ] v2.0.0
61 INFO: [Kivy        ] v2.0.0
[INFO   ] [Kivy        ] Installed at "/home/kewal/.local/lib/python3.8/site-packages/kivy/__init__.py"
61 INFO: [Kivy        ] Installed at "/home/kewal/.local/lib/python3.8/site-packages/kivy/__init__.py"
[INFO   ] [Python      ] v3.8.10 (default, Mar 15 2022, 12:22:08) 
[GCC 9.4.0]
61 INFO: [Python      ] v3.8.10 (default, Mar 15 2022, 12:22:08) 
[GCC 9.4.0]
[INFO   ] [Python      ] Interpreter at "/usr/bin/python3"
62 INFO: [Python      ] Interpreter at "/usr/bin/python3"
[INFO   ] [KivyMD      ] 1.0.0.dev0, git-9fe9417, 2022-01-11 (installed at "/home/kewal/.local/lib/python3.8/site-packages/kivymd/__init__.py")
62 INFO: [KivyMD      ] 1.0.0.dev0, git-9fe9417, 2022-01-11 (installed at "/home/kewal/.local/lib/python3.8/site-packages/kivymd/__init__.py")
[INFO   ] [Factory     ] 186 symbols loaded
62 INFO: [Factory     ] 186 symbols loaded
[INFO   ] [Image       ] Providers: img_tex, img_dds, img_sdl2, img_pil (img_ffpyplayer ignored)
92 INFO: [Image       ] Providers: img_tex, img_dds, img_sdl2, img_pil (img_ffpyplayer ignored)
[INFO   ] [Text        ] Provider: sdl2
95 INFO: [Text        ] Provider: sdl2
354 INFO: checking Analysis
652 INFO: checking PYZ
673 INFO: checking PKG
673 INFO: Bootloader /home/kewal/.local/lib/python3.8/site-packages/PyInstaller/bootloader/Linux-64bit-intel/run
673 INFO: checking EXE
674 INFO: Building because icon changed
674 INFO: Building EXE from EXE-00.toc
675 INFO: Copying bootloader EXE to /home/kewal/mowito_ws/arm_ws/src/order_control/scripts/build/operator_dashboard/operator_dashboard
675 INFO: Appending PKG archive to custom ELF section in EXE
693 INFO: Building EXE from EXE-00.toc completed successfully.
711 INFO: checking COLLECT
WARNING: The output directory "/home/kewal/mowito_ws/arm_ws/src/order_control/scripts/dist/operator_dashboard" and ALL ITS CONTENTS will be REMOVED! Continue? (y/N)y
On your own risk, you can use the option `--noconfirm` to get rid of this question.
3034 INFO: Removing dir /home/kewal/mowito_ws/arm_ws/src/order_control/scripts/dist/operator_dashboard
3487 INFO: Building COLLECT COLLECT-00.toc
5413 INFO: Building COLLECT COLLECT-00.toc completed successfully.



**But it does not relect the icon change in the executable**

**Reply by bwoodsend** [reply]
Executables don't have icons on Linux. You get `.desktop` files instead but producing those is outside of PyInstaller's territory.

**Reply by hakunaMatataHub** [reply]
Oh thanks.
How do i create .desktop file for my pyinstaller executable then ?

**Reply by bwoodsend** [reply]
https://www.baeldung.com/linux/desktop-entry-files

**Comment by hakunaMatataHub** [comment]
Thank you. I got it.

On Sat, Jul 16, 2022 at 8:51 PM Brénainn Woodsend ***@***.***>
wrote:

> https://www.baeldung.com/linux/desktop-entry-files
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/pyinstaller/discussions/6948#discussioncomment-3162517>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMGTBC6IXIULWSP7MEKP4M3VULHONANCNFSM53VMUK3Q>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***
> .com>
>


-- 

Kewal Bajaj
Indian Institute of Technology Kharagpur
[image: Image result for iit kgp]
Contact: +91 7978171236

**Comment by pskopnik** [comment]
Relying on `--collect-submodules` is probably simpler than specifying all individual hidden imports manually.

In my case:

```
$ pyinstaller --collect-submodules rclpy --collect-submodules rcl_interfaces --collect-submodules builtin_interfaces --collect-submodules std_msgs main.py
```

Which modules are imported will depend on the specific app.
Re-trying until all required modules are included may be preferable over including all modules included with ROS - again depending on your use case.

**Reply by pskopnik** [reply]
ROS would be a great candidate for creating hooks shipped with Pyinstaller. However, at this time I don't work with ROS enough to tackle all the various packages which may be imported and their hidden dependencies.
